drv/lpc55-sprot-server/src/main.rs

```diff
@@ -220,132 +216,136 @@ impl Io {
-                    && rx_msg.len() > 0
-                    && rx_msg.protocol() != Some(Protocol::Ignore)
-                {
-                    // This error matters
-                    err = Some(IoError::Flow);
-                }
-            }
+        ringbuf_entry!(Trace::ReceivedBytes(bytes_received));
 
-            // Are we done?
-            if csn_deasserted {
-                self.spi.ssd_clear();
+        // We don't bother sending bytes when receiving. So we must clear
+        // the underrun error condition before we handle a reply.
+        self.spi.txerr_clear();
 
-                if rx_msg.is_empty() {
-                    // This was a CSn pulse
-                    self.stats.csn_pulses =
-                        self.stats.csn_pulses.wrapping_add(1);
-                    err = Some(IoError::Flush);
-                }
+        Ok(bytes_received)
+    }
 
-                break;
-            }
+    fn reply(&mut self, tx_buf: &[u8]) {
+        ringbuf_entry!(Trace::ReplyLen(tx_buf.len()));
+
+        let mut tx = tx_buf.iter();
+
+        // Fill in the fifo before we assert ROT_IRQ
+        // We assert ROT_IRQ in `wait_for_csn_asserted` so we can
+        // put it after the `sys_irq_control` syscall to minimize time taken to
+        // process a request.
+        self.spi.drain_tx();
+        while self.spi.can_tx() {
+            let b = tx.next().copied().unwrap_or(0);
+            self.spi.send_u8(b);
+        }
+
+        self.wait_for_csn_asserted(true);
 
-            // Read the CSn flag *after* checking if CSn is deasserted.
-            //
-            // This ordering allows us to loop one more time to exchange data
-            // and empty our fifos. Putting the check at the end also minimizes
-            // the time to access the fifos the first time through the loop.
-            // The goal here is to reduce overrun/underrun after already
-            // waiting for the CSn asserted interrupt to fire.
-            csn_deasserted = self.spi.ssd();
+        let mut bytes_sent = 0;
+        while !self.spi.ssd() {
+            while self.spi.can_tx() {
+                bytes_sent += 1;
+                let b = tx.next().copied().unwrap_or(0);
+                self.spi.send_u8(b);
+            }
         }
 
-        if too_many_bytes_received {
-            self.stats.rx_protocol_error_too_many_bytes =
-                self.stats.rx_protocol_error_too_many_bytes.wrapping_add(1);
+        self.spi.ssd_clear();
+
+        // We clocked out at least an existing byte in the fifo, as we fill it on entry to
+        // this function.
+        if bytes_sent == 0 || self.spi.can_tx() {
+            // This was a CSn pulse
+            self.stats.csn_pulses = self.stats.csn_pulses.wrapping_add(1);
+        } else {
+            self.check_for_tx_error();
         }
 
-        err.map_or(Ok(()), |e| Err(e))
+        ringbuf_entry!(Trace::SentBytes(bytes_sent));
+        self.prepare_for_next_request();
```

Comment:
I'd suggest inlining `prepare_for_next_request` into this function – it's only ever called from here and does very specific things, but has a tempting-enough name that I'd imagine someone trying to call it in the future.

Reply:
Done.